### ISSUE TITLE
Add ability to toggle critic portion of the network

### ISSUE THREAD

**Issue by SBeairsto** [climagination/ClimatExML#32]
### ✨ Feature Request: Toggle Critic On/Off via Config

**Summary:**  
Add a simple config-based flag to enable or disable the Critic component of the WGAN-GP model. This would allow users to easily switch between full WGAN-GP mode and a simpler generator-only (pure CNN) mode.

---

**Motivation:**  
When sharing this codebase with new collaborators or users, the most common first step is to run the generator-only version of the model to get familiar with the architecture and data. Right now, doing this requires commenting out multiple lines of code manually.

Having a config flag like `include_critic: true/false` would:
- Make the training behavior easier to control without modifying source code
- Reduce friction for first-time users and facilitate experimentation
- Align with the existing config-driven design pattern already used in the repo

---

**Proposed Changes:**
- Add a new entry to `config.yaml`:
  ```yaml
  include_critic: false

- In SuperResolutionWGANGP, wrap critic-related logic in a conditional:
         - e.g., only define `self.C`, compute the gradient penalty, toggle optimizers, etc. if `include_critic` is `True`

- During `training_step` and `validation_step`, skip critic-related calculations and logging when disabled

- Optionally suppress or stub out critic checkpoint saving when not used

---

**Outcome:**
This would make it seamless to switch between full GAN training and generator-only CNN training for debugging, experimentation, and onboarding. It’s also a step toward modularizing the architecture for broader use cases.

Let me know if this enhancement makes sense — happy to help implement or test it.

**Comment by nannau** [comment]
I fully agree with the premise that this would be very useful! I have some software engineering thoughts on how to best approach this.

Sometimes, even the simplest of flags can severely bloat a project and get in the way of making it modular because of baked in assumptions (e.g. tracking the wasserstein loss in the loss logging functions requires different behaviour with/without the critic). This is the main reason I haven't implemented this as a flag in `ClimatExML`. A concrete example is a series of if/else statements littered throughout the code that control specific behaviour. It's hard to maintain such code, and the added logic can be very confusing especially as more flags are added. 

However, since there are many more other interesting ways to run training algorithms in `ClimatExML` besides a GAN or even just a straight MSE/MLE super resolution task (i.e. without the critic), we should consider modularizing the core training loop so that a user can select how they want to train the networks (which are also modular!). I can see this being useful for when things like diffusion models are introduced, too. A user of `ClimatExML` should conceivably be able to use any form of compatible training algorithm, provided they implement it correctly, without them requiring beforehand knowledge of how one specific implementation works (e.g. `wgan_gp.py`)

My suggestion to approach this is to approach it like this:

1. Create an abstract base class that implement the core _required_ functions from `wgan_gp.py` _as abstract methods_. Some examples that come to mind are things like `losses()`, `training/validation_step()`, `configure_optimizers()` etc.
2. Factor out functions in `wgan_gp.py` that can be used in _any_ sufficiently complicated super resolution training loop like `go_downnhill()`. 
3. Create a new file `gan.py` and `cnn.py` (or some other good name) that are concrete and fully implement specific versions of `losses()`, `training/validation_step()` etc. 
4. Carefully consider whether functions (e.g.`losses()`) can be made general enough to work for any training setup or whether there are specific things that require it's specific implementation in the concrete class instances. 
5. Allow for a user to select either `gan` or `cnn` in the configs and then instantiate the concrete classes from there.

Hopefully that would reduce coupling in the code and help keep it modular. Feel free to ask questions because I'm throwing a lot of software engineering ideas out there.

**Comment by SBeairsto** [comment]
Beautiful, thanks Nic! This should be plenty to go on, I'll get cracking on it and let you know if I run into any roadblocks!

**Comment by nannau** [comment]
Thanks, @SBeairsto! This issue will be a good place to pontificate on high-level design ideas. Let me know how the design works/feels for you. What I've proposed is basically how lightning works -- they call these special functions "hooks". Basically, the only difference is that we further constrain these hooks so that they fit our problem space best.
